.github/workflows/smoke-test-daily.yml

```diff
@@ -17,22 +17,15 @@ concurrency:
 permissions: {}
 
 jobs:
-    e2e-tests:
-        name: E2E tests on nightly build
+    api-tests:
+        name: API tests on nightly build
         runs-on: ubuntu-20.04
         permissions:
           contents: read
+        if: success() || failure()
```

Comment:
Good shout @rodelgc!

Question:
`if: success() || failure()` will run the action in the step whenever the previous step had been successful or failed so makes me wonder if this was necessary

By default, GitHub Actions steps are set up to execute regardless of the success or failure of the preceding step. This means that if a preceding step fails, the next step will still run.

So - I am wondering why we need the `if: success() || failure()` ?

Reply:
Great question, as I was a bit confused myself at first on how steps and job dependencies worked in GitHub Actions.

> By default, GitHub Actions steps are set up to execute regardless of the success or failure of the preceding step

`if` conditions can be placed on the step level, and at the job level.

If we're talking about "steps", by default a step inside a job has an implicit `if: success()` and will run only if [none of the previous steps have failed or been canceled](https://docs.github.com/en/actions/learn-github-actions/expressions#success). Let's say, if:
 - `step 1` failed ❌ 
 - and `step 2` passed ✔️
 
Then `step 3` by default, will be **skipped** because `step 1` failed, even though the preceding `step 2` passed. As long as any of the earlier steps failed, whether they're the immediately preceding or a way earlier step (within the same job), then the current step will be skipped.

`step 3`  will **run** if it has `if: failure()`.

But giving it with just `if: failure()` alone will cause a different issue: `step 3` will not run if both `step 1` and `step 2` passed ✔️. That's the limitation of `if: failure()` -- it only runs `step 3` if there's at least one previous step that failed. If none of the previous steps failed, then `if: failure()` returns `false`, thereby preventing `step 3` from running.

If we want `step 3` to run when either:
- both `step 1` and `step 2` passed ✔️ 
- at least one of the earlier steps failed ❌ 
Then that's the time when we'll need to give it an `if: success() || failure()`. 

A good example is the [Generate API test report](https://github.com/woocommerce/woocommerce/blob/7cf15992086dbbedd1c74c310b9869948ce77cbf/.github/workflows/smoke-test-daily.yml#L49-L50) step. Notice that it has `if: failure() || success()` because we want it to run (generate the report) whether the preceding [Run API tests](https://github.com/woocommerce/woocommerce/blob/7cf15992086dbbedd1c74c310b9869948ce77cbf/.github/workflows/smoke-test-daily.yml#L40) step passed or failed. Without `if: failure() || success()`, "Generate API test report" will not run if the API tests failed.

----

Now if we're talking about "jobs", then the same default behavior in "steps" would apply, but only on jobs depending on each other. For example, in the image below, `Job 4` still ran because it only depends on `Job 3`. It ignores the failure in `Job 2` because it doesn't depend on it. As long as `Job 3` passes, then `Job 4` would run.

<img width="762" alt="rKDyeS2XTk" src="https://user-images.githubusercontent.com/4509348/214770926-0e85e5ac-6a13-4556-a1e2-7c53f755e97d.png">

In the case of the `api-tests` job, since it's  now at the beginning of the workflow, it does not depend on any previous job. So we don't need an `if: success() || failure()` anymore. However, we still need the `if: success() || failure()` condition in the [`trunk-results`](https://github.com/woocommerce/woocommerce/blob/7cf15992086dbbedd1c74c310b9869948ce77cbf/.github/workflows/smoke-test-daily.yml#L252) job so that even if the API test failed, we would still be able to publish the report on it.